googlemock/include/gmock/gmock-generated-nice-strict.h.pump

```diff
@@ -94,38 +94,49 @@ class $clazz : public MockClass {
   // We don't factor out the constructor body to a common method, as
   // we have to avoid a possible clash with members of MockClass.
   $clazz() {
-    ::testing::Mock::$method(
-        internal::ImplicitCast_<MockClass*>(this));
+    ::testing::Mock::$method(this);
   }
 
   // C++ doesn't (yet) allow inheritance of constructors, so we have
   // to define it for each arity.
   template <typename A1>
   explicit $clazz(const A1& a1) : MockClass(a1) {
-    ::testing::Mock::$method(
-        internal::ImplicitCast_<MockClass*>(this));
+    ::testing::Mock::$method(this);
   }
 
 $range i 2..n
 $for i [[
 $range j 1..i
   template <$for j, [[typename A$j]]>
   $clazz($for j, [[const A$j& a$j]]) : MockClass($for j, [[a$j]]) {
-    ::testing::Mock::$method(
-        internal::ImplicitCast_<MockClass*>(this));
+    ::testing::Mock::$method(this);
   }
 
 
 ]]
   virtual ~$clazz() {
-    ::testing::Mock::UnregisterCallReaction(
-        internal::ImplicitCast_<MockClass*>(this));
+    ::testing::Mock::UnregisterCallReaction(this);
   }
 
  private:
   GTEST_DISALLOW_COPY_AND_ASSIGN_($clazz);
 };
 
+namespace internal {
+
+template <typename MockClass>
+struct MockTraits<$clazz<MockClass> > {
+  static void* AddressOf($clazz<MockClass>* mock_obj) {
```

Comment:
We shouldn't need `void*`.
I was thinking of something less intrusive.

    template <typename M>
    struct UnwrappedMockType { using type = M; };

    // Each of these specializations would follow their class template definitions.
    template <typename M>
    struct UnwrappedMockType<StrictMock<M>> { using type = M; };
    template <typename M>
    struct UnwrappedMockType<NiceMock<M>> { using type = M; };
    template <typename M>
    struct UnwrappedMockType<NaggyMock<M>> { using type = M; };

You'd use an ImpliciitCast to `UnwrappedMockType<M>::type*` where needed.